File: 3DEngine/Delaunay/Triangle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Delaunay
{
    /**
     * 三角形クラス
     *
     * @author tercel
     */
    public class Triangle
    {
        public Point3d v1, v2, v3;
        public Triangle(Point3d v1, Point3d v2, Point3d v3)
        {
            this.v1 = v1;
            this.v2 = v2;
            this.v3 = v3;
        }

        // Get the Normal Vector
        // It is assumed that vertices are in counterclockwise order
        public Point3d GetNormal()
        {
            Point3d edge1 = new Point3d(v2.X - v1.X, v2.Y - v1.Y, v2.Z - v1.Z);
            Point3d edge2 = new Point3d(v3.X - v1.X, v3.Y - v1.Y, v3.Z - v1.Z);

            // x
            Point3d normal = edge1.Cross(edge2);
            normal.Normalize();
            return normal;
        }

        // Reverse the order of the vertices, to flip the surface
        public void TurnBack()
        {
            Point3d tmp = this.v3;
            this.v3 = this.v1;
            this.v1 = tmp;
        }

        // Get a list of line segments
        public Line[] GetLines()
        {
            Line[] l = {
                        new Line(v1, v2),
                        new Line(v2, v3),
                        new Line(v3, v1)
                    };
            return l;
        }

        public Point3d[] Vertices()
        {
            return new Point3d[] {v1,v2,v3};
        }

        public void SetVertices(Point3d v1,Point3d v2,Point3d v3)
        {
            this.v1 = v1;
            this.v2 = v2;
            this.v3 = v3;
        }

        public override bool Equals(object obj) {
            if(obj is Triangle) {
                Triangle t = (Triangle)obj;
                Line[] lines1 = this.GetLines();
                Line[] lines2 = t.GetLines();

                int cnt = 0;
                for(int i = 0; i < lines1.Count(); i++) {
                    for(int j = 0; j < lines2.Count(); j++) {
                        if(lines1[i].Equals(lines2[j]))
                            cnt++;
                    }
                }

                return (cnt == 3);
            } else {
                return false;
            }
        }

        public override int GetHashCode() {
            return base.GetHashCode();
        }
    }
}

File: 3DEngine/Classes/Delaunay/Triangle.vb
Namespace Delaunay
    Public Class Triangle
        Implements IEquatable(Of Triangle)

        Private p1 As Point3d
        Private p2 As Point3d
        Private p3 As Point3d

        Public Sub New(p1 As Point3d, p2 As Point3d, p3 As Point3d)
            Me.p1 = p1
            Me.p2 = p2
            Me.p3 = p3
        End Sub

        Public ReadOnly Property Normal As Point3d
            Get
                Dim edge1 As Point3d = p2 - p1
                Dim edge2 As Point3d = p3 - p1

                Return edge1.Cross(edge2).Normalized()
            End Get
        End Property

        Public Sub Flip()
            Dim tmp As Point3d = New Point3d(p3)
            p3 = New Point3d(p1)
            p1 = tmp
        End Sub

        Public ReadOnly Property Vertices As Point3d()
            Get
                Return New Point3d() {p1, p2, p3}
            End Get
        End Property

        Public ReadOnly Property Lines As Line()
            Get
                Return {New Line(p1, p2),
                        New Line(p2, p3),
                        New Line(p3, p1)}
            End Get
        End Property

        Public Shared Operator =(t1 As Triangle, t2 As Triangle) As Boolean
            Dim lines1() As Line = t1.Lines()
            Dim lines2() As Line = t2.Lines()

            Dim counter As Integer = 0
            For i As Integer = 0 To lines1.Length - 1
                For j As Integer = 0 To lines2.Length - 1
                    If lines1(i) = lines2(j) Then counter += 1
                Next
            Next

            Return counter = 3
        End Operator

        Public Shared Operator <>(t1 As Triangle, t2 As Triangle) As Boolean
            Return Not (t1 = t2)
        End Operator

        Public Shadows Function Equals(other As Triangle) As Boolean Implements IEquatable(Of Triangle).Equals
            Return Me = other
        End Function
    End Class
End Namespace
